### PR TITLE
with controllers

### PR DESCRIPTION
**Pull request by sv053**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -33,15 +37,16 @@
                 </exclusion>
             </exclusions>
         </dependency>
+
         <dependency>
             <groupId>org.springframework.boot</groupId>
-            <artifactId>spring-boot-starter-log4j2</artifactId>
```
Comment (AlexeyShutov) [inline pom.xml:38]: why?

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+import java.util.List;
+import java.util.Optional;
+
+@RestController
+@RequestMapping("/games")
+public class GameController {
+
+    private final GameRepository gameRepository;
+    private final Logger logger = LoggerFactory.getLogger(getClass());
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:18]: loggers are usually static

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+import java.util.Optional;
+
+@RestController
+@RequestMapping("/games")
+public class GameController {
+
+    private final GameRepository gameRepository;
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:17]: We do you even need services in this scenario?

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    public GameController(GameRepository gameRepository) {
+        this.gameRepository = gameRepository;
+    }
+
+    @PostMapping("/createGames")
+    public List<Game> createGames(@RequestBody List<Game> gamesToSave){
+        gameRepository.createAll(gamesToSave);
+        gamesToSave.stream().forEach(u -> logger.info(u.toString()));
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:27]: Don't produce excessive logging that doesn't bring any useful info

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        return gamesToSave;
+    }
+
+    @PostMapping("/findGame")
+    public Game findGameById(@RequestParam Integer id) {
+        Optional<Game> game = gameRepository.findById(id);
+        if(game.isPresent()){
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:34]: This is just broken

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+public class GameController {
+
+    private final GameRepository gameRepository;
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    public GameController(GameRepository gameRepository) {
+        this.gameRepository = gameRepository;
+    }
+
+    @PostMapping("/createGames")
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:24]: You don't need to specify actions with names, you have verbs in REST for this. The only exclusions are the endpoints that cannot be described by the rest verbs.

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    public List<Game> createGames(@RequestBody List<Game> gamesToSave){
+        gameRepository.createAll(gamesToSave);
+        gamesToSave.stream().forEach(u -> logger.info(u.toString()));
+        return gamesToSave;
+    }
+
+    @PostMapping("/findGame")
+    public Game findGameById(@RequestParam Integer id) {
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:32]: Instead of parameter - the ID could be a part of the path

File: src/main/java/com/gamesage/store/controller/UserController.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    public UserController(UserRepository userRepository) {
+        this.userRepository = userRepository;
+    }
+
+    @PostMapping("/findUser")
+    public User findUserById(@RequestParam Integer id) {
+        User user = null;
+        if(userRepository.findById(id).isPresent()){
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/UserController.java:25]: 2 queries to repo just to check if user exists?

File: src/main/java/com/gamesage/store/domain/model/Game.java [hunk 9]
```diff
@@ -9,10 +9,6 @@
     private final String name;
     private final BigDecimal price;
 
-    public Game(String name, BigDecimal price) {
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/domain/model/Game.java:12]: That was a bad move, it's always nice to have a reloaded stuff to indicate parameters with default values

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+    private final GameRepository gameRepository;
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    public GameController(GameRepository gameRepository) {
+        this.gameRepository = gameRepository;
+    }
+
+    @PostMapping("/createGames")
+    public List<Game> createGames(@RequestBody List<Game> gamesToSave){
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:25]: Formatting

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
+package com.gamesage.store.controller;
+
+import com.gamesage.store.domain.data.sample.SampleData;
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:3]: redundant

File: src/main/java/com/gamesage/store/domain/repository/UserRepository.java [hunk 12]
```diff
@@ -27,6 +27,11 @@ public UserRepository() {
         return users;
     }
 
+    public void createUser(User user){
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/domain/repository/UserRepository.java:30]: 1. Looks like this repo doesn't assign any IDs to users
2. Create should return the results

**Review by AlexeyShutov — CHANGES_REQUESTED** [review]
Changes are proposed/indicated in the comments

File: src/main/java/com/gamesage/store/domain/repository/Repository.java [hunk 13]
```diff
@@ -1,14 +1,13 @@
...
 import java.util.Optional;
 
 public interface Repository<T, K> {
 
     Optional<T> findById(K id);
-
-    List<T> getAll();
-
-    void createAll(List<T> items);
+    List<T> find();
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/domain/repository/Repository.java:10]: We already had one long discussion about `findBy`, now this one is even more confusing

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    @PostMapping("/create")
+    public List<Game> createGames(@RequestBody List<Game> gamesToSave){
+        gameService.createAll(gamesToSave);
+        GameController.logger.info(gamesToSave.size() + " games were added");
+        return gameService.findAll();
+    }
+
+    @PostMapping("/read/{id}")
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:29]: `POST` for fetching a resource?

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+public class GameController {
+
+    private final GameService gameService;
+    private static final Logger logger = LoggerFactory.getLogger(GameController.class);
+
+    public GameController(GameService gameService) {
+        this.gameService = gameService;
+    }
+
+    @PostMapping("/create")
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:22]: These are still redundant, the REST verbs are already set

File: src/main/java/com/gamesage/store/controller/UserController.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/users")
+public class UserController {
+
+    private final UserService userService;
+    private static final Logger logger = LoggerFactory.getLogger(UserController.class);
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/UserController.java:16]: Wrong order

File: src/main/java/com/gamesage/store/controller/UserController.java [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    @GetMapping("/read")
+    public List<User> findAllUsers() {
+        List<User> users = userService.findAll();
+        UserController.logger.info("There are " + users.size() + " users");
+        return users;
+    }
+
+    @PostMapping("/createOne")
+    public User createUser(@RequestBody User user){
+        userService.createAll(List.of(user));
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/UserController.java:38]: Why not to add a method for a single user?

File: src/main/java/com/gamesage/store/controller/UserController.java [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(UserController.class);
+
+    public UserController(UserService userService) {
+        this.userService = userService;
+    }
+
+    @PostMapping("/read/{id}")
+    public User findUserById(@PathVariable Integer id) {
+        User user = userService.findById(id);
+        UserController.logger.info("User: " + user);
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/UserController.java:25]: full names again

File: src/main/java/com/gamesage/store/controller/UserController.java [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+        return user;
+    }
+
+    @GetMapping("/read")
+    public List<User> findAllUsers() {
+        List<User> users = userService.findAll();
+        UserController.logger.info("There are " + users.size() + " users");
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/UserController.java:32]: Such logs would kill your machine in prod

File: src/main/java/com/gamesage/store/domain/model/Game.java [hunk 20]
```diff
@@ -1,15 +1,23 @@
...
-public class Game {
+@JsonDeserialize(as = Game.class)
+public class Game{
 
     private Integer id;
     private final String name;
     private final BigDecimal price;
 
-    public Game(String name, BigDecimal price) {
+    public Game(){
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/domain/model/Game.java:15]: There is a more graceful way to solve the deserialization problem

File: src/main/java/com/gamesage/store/controller/UserController.java (hunk 19 above)
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/UserController.java:32]: What is the use of this log?

File: src/main/java/com/gamesage/store/domain/data/sample/SampleData.java [hunk 21]
```diff
@@ -16,14 +16,14 @@
...
         GAMES = List.of(
-                new Game("THE_WITCHER", BigDecimal.valueOf(17.28d)),
-                new Game("GRAND_THEFT_AUTO", RandomBigDecimal.getAndFormatRandomBigDecimal()),
-                new Game("RED_DEAD_REDEMPTION", RandomBigDecimal.getAndFormatRandomBigDecimal()),
-                new Game("SKYRIM", BigDecimal.valueOf(87.88d)),
-                new Game("FORZA_HORIZON", RandomBigDecimal.getAndFormatRandomBigDecimal()),
-                new Game("DOOM_ETERNAL", BigDecimal.valueOf(635.48d)),
-                new Game("ASSASSIN_S_CREED", RandomBigDecimal.getAndFormatRandomBigDecimal()),
-                new Game("A_NEW_ONE", RandomBigDecimal.getAndFormatRandomBigDecimal()));
+                new Game(null,"THE_WITCHER", BigDecimal.valueOf(17.28d)),
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/domain/data/sample/SampleData.java:19]: No changed back

File: src/main/java/com/gamesage/store/controller/UserController.java (hunk 16 above)
Comment (sv053) [reply src/main/java/com/gamesage/store/controller/UserController.java:16]: PSF pattern?

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+@RequestMapping("/games")
+public class GameController {
+
+    private final GameService gameService;
+
+    public GameController(GameService gameService) {
+        this.gameService = gameService;
+    }
+
+    @PostMapping("/")
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:19]: Why adding this rudiment to URI?

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+    public List<Game> createGames(@RequestBody List<Game> gamesToSave){
+        gameService.createAll(gamesToSave);
+        return gameService.findAll();
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:22]: Why anyone would want to see all the database after creating a bunch of games?

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    @GetMapping("/{id}")
+    public Game findGameById(@PathVariable Integer id) {
+        Game game = gameService.findById(id);
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:27]: Redundant var

File: src/main/java/com/gamesage/store/controller/UserController.java [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+        return user;
+    }
+
+    @GetMapping("/")
+    public List<User> findAllUsers() {
+        List<User> users = userService.findAll();
+        return users;
+    }
+
+    @PostMapping("/userslist")
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/UserController.java:31]: I really don't see a use case for creating multiple users at once

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+        return game;
+    }
+
+    @GetMapping("/")
+    public List<Game> findGames(){
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:32]: find what games?

File: src/main/java/com/gamesage/store/domain/model/Game.java [hunk 27]
```diff
@@ -1,19 +1,27 @@
...
 import java.math.BigDecimal;
 import java.util.Objects;
 
-public class Game {
+public class Game{
 
     private Integer id;
     private final String name;
     private final BigDecimal price;
+    static final long serialVersionUID = 1L;
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/domain/model/Game.java:15]: Why do you need serialVersionUID?

File: src/test/java/com/gamesage/store/domain/repository/GameRepositoryTest.java [hunk 28]
```diff
@@ -37,10 +37,10 @@ void createAll() {
         List<Game> games = List.of(
                 new Game("addedGame1", null),
                 new Game("addedGame2", null));
-        repository.createAll(games);
+        repository.create(games);
 
-        assertEquals(games.size(), repository.getAll().size());
-        assertTrue(repository.getAll().containsAll(games));
+        assertEquals(games.size(), repository.findAll().size());
```
Comment (AlexeyShutov) [inline src/test/java/com/gamesage/store/domain/repository/GameRepositoryTest.java:42]: Why making multiple requests?

File: src/main/java/com/gamesage/store/controller/UserController.java [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+    public List<User> createUsers(@RequestBody List<User> users){
+        userService.create(users);
+        return users;
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/UserController.java:34]: A client doesn't want to see the users he sends, but created users

**Review by AlexeyShutov — CHANGES_REQUESTED** [review]
1. Design issues mentioned in the comments
2. Code is not formatted in many places
3. Unused imports in multiple files

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 30]
```diff
@@ -0,0 +1,34 @@
...
+@RequestMapping("/games")
+public class GameController {
+
+    private final GameService gameService;
+
+    public GameController(GameService gameService) {
+        this.gameService = gameService;
+    }
+
+    @PostMapping("")
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:19]: An empty path is redundant

File: src/main/java/com/gamesage/store/domain/model/Game.java [hunk 31]
```diff
@@ -1,19 +1,25 @@
...
+
 import java.math.BigDecimal;
 import java.util.Objects;
 
 public class Game {
 
-    private Integer id;
     private final String name;
     private final BigDecimal price;
+    private Integer id;
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/domain/model/Game.java:13]: ID usually is the first field

File: src/main/java/com/gamesage/store/controller/UserController.java [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+
+import com.gamesage.store.domain.model.User;
+import com.gamesage.store.service.UserService;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/users")
+public class UserController {
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/UserController.java:11]: Requested changes are not applied to the controller's methods

File: src/main/java/com/gamesage/store/domain/repository/GameRepository.java [hunk 33]
```diff
@@ -20,27 +20,38 @@ public GameRepository() {
 
     @Override
     public Optional<Game> findById(Integer key) {
-        return Optional.ofNullable(allGamesById.get(key));
+        return Optional
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/domain/repository/GameRepository.java:23]: The first operation may stay on the same line

File: src/main/java/com/gamesage/store/service/GameService.java [hunk 34]
```diff
@@ -3,24 +3,34 @@
...
-import com.gamesage.store.domain.repository.Repository;
+import com.gamesage.store.domain.repository.AddManyRepository;
 import org.springframework.stereotype.Service;
 
 import java.math.BigDecimal;
+import java.util.List;
 
 @Service
-public class GameService {
+public class GameService{
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/service/GameService.java:13]: ->

File: src/main/java/com/gamesage/store/domain/model/Game.java [hunk 35]
```diff
@@ -1,19 +1,25 @@
 package com.gamesage.store.domain.model;
 
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
 import java.math.BigDecimal;
 import java.util.Objects;
 
 public class Game {
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/domain/model/Game.java:9]: You're missing a separator in `toString` between `id` and `name`

File: src/test/java/com/gamesage/store/domain/repository/GameRepositoryTest.java [hunk 36]
```diff
@@ -37,10 +37,9 @@ void createAll() {
         List<Game> games = List.of(
                 new Game("addedGame1", null),
                 new Game("addedGame2", null));
-        repository.createAll(games);
+        repository.create(games);
 
-        assertEquals(games.size(), repository.getAll().size());
```
Comment (AlexeyShutov) [inline src/test/java/com/gamesage/store/domain/repository/GameRepositoryTest.java:42]: I didn't tell that you don't need to assert the size, because `containsAll` tests only the sub-list, and there are no guarantees that no duplicates are created

File: src/main/java/com/gamesage/store/domain/repository/AddManyRepository.java [hunk 37]
```diff
@@ -0,0 +1,9 @@
+package com.gamesage.store.domain.repository;
+
+import java.util.List;
+
+public interface AddManyRepository<T, K> extends Repository<T, K>{
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/domain/repository/AddManyRepository.java:5]: ` `

Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/domain/repository/AddManyRepository.java:5]: I'm not sure that `add` gives a good context here, you have a limited functionality that just creates new entities and doesn't merge the changes

File: src/main/java/com/gamesage/store/controller/GameController.java [hunk 38]
```diff
@@ -0,0 +1,34 @@
...
+
+import com.gamesage.store.domain.model.Game;
+import com.gamesage.store.service.GameService;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/games")
+public class GameController {
```
Comment (AlexeyShutov) [inline src/main/java/com/gamesage/store/controller/GameController.java:11]: Follow similar order for methods like in repo/services, it is easier to find everything when a controller grows

**Review by AlexeyShutov — APPROVED** [review]
